### PR TITLE
Add preprocess to synthesizers

### LINKED ISSUE #1018: Add preprocess method to synthesizers
### Problem Description

_As a user, it would be useful to transform my data without having to fit the model._

In order to allow users to separately transform and model data, the functionality should be split into separate methods. The first is the `preprocess` method.

### Expected behavior

- Add `preprocess` method to `BaseSynthesizer` class.
- Parameters:
    - data (`pandas.DataFrame`): The data to transform.
- The method should do the following steps:
    1. Validates the data against the metadata (call `synthesizer.validate`)
    2. Transforms the data using constraints
    3. (if not already done) Creates HyperTransformer(s) + assign transformers 
    4. (if not already fit) Fits HyperTransformers(s)
    5. Transforms data using the HyperTransformer(s)
    6. Return transformed data
- Note that for `CTGAN`, `TVAE` and `CopulaGAN`, booleans and categorical columns should not be transformed

#### Warnings
- If the model has already been fit, raise the following warning:
`Warning: This model has already been fit. To use the new preprocessed data, please refit the model using 'fit' or 'fit_processed_data'`

### Additional context

- We may want to add a parameter to the `DataProcessor` that let's us override defaults for `sdtypes` to account for the `CTGAN`, `TVAE` and `CopulaGAN` cases. This is similar to what we do in the current SDV version. We could also directly update the transformers to use in the `HyperTransformer`.
- If the `HyperTransformer` has not already been created, this method can call the `assign_transformers` method added in #1020

### PR DESCRIPTION
**Pull request by pvk-developer**
Resolves #1018

### CODE REVIEW

**Review by amontanez24 — CHANGES_REQUESTED** [review]

File: tests/unit/data_processing/test_data_processor.py [hunk 1]
```diff
@@ -441,6 +441,26 @@ def test__fit_transform_constraints_missing_columns_error(self, log_mock):
...
+            'categorical': 'labelencoder',
+            'numerical': 'float',
+            'boolean': None
+        }
+
+        # Run
+        DataProcessor._update_transformers_by_sdtypes(instance, 'categorical', None)
+
+        # Assert
+        instance._transformers_by_sdtype == {
```
Comment (fealho) [inline tests/unit/data_processing/test_data_processor.py:458]: add assert

File: tests/unit/single_table/test_base.py [hunk 2]
```diff
@@ -75,6 +75,46 @@ def test_get_metadata(self, mock_data_processor):
         # Assert
         assert result == metadata
 
+    @patch('sdv.single_table.base.warnings')
+    def test_preprocess(self, mock_warnings):
+        """Test that ``preprocess`` calls the ``validate`` then fits and returns transformed data.
+
+        Before the preprocessing is occurs we check that the synthesizer has
```
Comment (fealho) [inline tests/unit/single_table/test_base.py:82]: typo: delete is

**Review by fealho — COMMENTED** [review]
Looks good! I'm just not sure whether you implemented the last comment in the issue (`If the HyperTransformer has not already been created, this method can call the assign_transformers method added in`)?

File: sdv/single_table/base.py [hunk 3]
```diff
@@ -276,3 +233,64 @@ def update_transformers(self, column_name_to_transformer):
...
+        self.validate(data)
+        if self._fitted:
+            warnings.warn(
+                'This model has already been fitted. To use the new preprocessed data, '
+                'please refit the model using `fit` or `fit_processed_data`.'
+            )
+
+        if self._model_sdtype_trnasformers is not None:
+            for sdtype, transformer in self._model_sdtype_trnasformers.items():
+                self._data_processor._update_transformers_by_sdtypes(sdtype, transformer)
```
[the remark this answers is not in the document]
Comment (pvk-developer) [reply sdv/single_table/base.py:262]: Yes we can move it to the `init`

File: tests/unit/single_table/test_base.py [hunk 4]
```diff
@@ -18,6 +18,23 @@
 
 class TestBaseSynthesizer:
 
+    def test__update_default_transformers(self):
+        """
```
Comment (fealho) [inline tests/unit/single_table/test_base.py:22]: Perhaps you could be slightly more descriptive in the docstring.

**Review by amontanez24 — APPROVED** [review]
:shipit:

**Review by fealho — APPROVED** [review]

**Pull request merged**